x-pack/plugins/apm/public/components/shared/Stacktrace/stacktraceUtils.ts

```diff
@@ -7,35 +7,35 @@
 import { get, isEmpty } from 'lodash';
 import { Stackframe } from '../../../../typings/APMDoc';
 
-export interface StackframeCollapsed extends Stackframe {
-  libraryFrame?: boolean;
-  stackframes?: Stackframe[];
+interface StackframesGroup {
+  isLibraryFrame: boolean;
+  stackframes: Stackframe[];
 }
 
-export function getCollapsedLibraryFrames(
+export function getGroupedStackframes(
   stackframes: Stackframe[]
-): StackframeCollapsed[] {
-  return stackframes.reduce((acc: any, stackframe: StackframeCollapsed) => {
-    if (!stackframe.libraryFrame) {
-      return [...acc, stackframe];
-    }
-
-    // current stackframe is library frame
-    const prevItem: StackframeCollapsed = acc[acc.length - 1];
-    if (!get(prevItem, 'libraryFrame')) {
-      return [...acc, { libraryFrame: true, stackframes: [stackframe] }];
-    }
-
-    return [
-      ...acc.slice(0, -1),
-      {
-        ...prevItem,
-        stackframes: prevItem.stackframes
-          ? [...prevItem.stackframes, stackframe]
-          : [stackframe]
-      }
-    ];
-  }, []);
+): StackframesGroup[] {
+  if (stackframes.length === 0) {
+    return [];
+  }
+
+  const isLibraryFrame = Boolean(stackframes[0].library_frame);
+
+  let i = 0;
+  while (
+    i < stackframes.length &&
+    isLibraryFrame === stackframes[i].library_frame &&
+    !stackframes[i].exclude_from_grouping
+  ) {
+    i++;
+  }
+
+  const stackFrameEndIndex = i || 1;
```

Comment:
minimum group size is 1, since 0 stackframes is not a valid or renderable group

Reply:
I don't think this code is super clear. Either extract it to a separate method like `getStackFrameEndIndex` and/or rewrite it without `while`.


Un-tested, but this is how I understand the current implementation
```js
const stackFrameEndIndex =
    stackframes.findIndex(
      stackframe =>
        isLibraryFrame === stackframe.library_frame &&
        !stackframe.exclude_from_grouping
    ) || 1;
```

Reply:
We can get rid of the recursiveness by using `reduce`:

```ts
interface StackframesGroup {
  isLibraryFrame: boolean;
  excludeFromGrouping: boolean | undefined;
  stackframes: Stackframe[];
}

export function getGroupedStackframes(
  stackframes: Stackframe[]
): StackframesGroup[] {
  return stackframes.reduce(
    (acc, stackframe) => {
      const prevGroup = last(acc);
      const isGroupable =
        prevGroup &&
        prevGroup.isLibraryFrame === stackframe.library_frame &&
        !prevGroup.excludeFromGrouping &&
        !stackframe.exclude_from_grouping;

      if (isGroupable) {
        prevGroup.stackframes.push(stackframe);
        return acc;
      }

      acc.push({
        isLibraryFrame: Boolean(stackframe.library_frame),
        excludeFromGrouping: stackframe.exclude_from_grouping,
        stackframes: [stackframe]
      });
      return acc;
    },
    [] as StackframesGroup[]
  );
}
```

I find the intend in the `reduce` a little clearer but that might just be me.

Reply:
... I just realise that you replaced the previous reducer with a recursive function - if you prefer that it's also totally fine.

Reply:
i kept the recursive algorithm, but used your first suggestion to replace the while loop with `findIndex` and externalized some logic in a help function.